doc/read-the-docs-site/adr/0005-plutus-tx-compiler-and-library.md

```diff
@@ -0,0 +1,205 @@
+But if make let-bindings strict by default, then the question is, how does one make a non-strict let-binding?
+
+Haskell supports lazy patterns with the `~` operator, and it can be used on let-bindings, e.g., `let ~x = rhs`. So one possibility is to make non-strict let-bindings this way.
+However, lazy pattern has no effect on simple pattern bindings (i.e., the pattern is a single variable), and it makes no difference in GHC Core, which means the Plutus Tx compiler can't actually see the `~`.
+It is technically possible to make `~` work but it would require adding a GHC parser plugin to capture `~` in GHC Core, among other things, adding a great deal of complexity.
+
+This is the reason why let-bindings have been non-strict by default in Plutus Tx, and creating strict let-bindings requires using bang patterns.
+
+However, we can do it in a different way.
+Although we can't modify Haskell's syntax, we can introduce functions to be handled specially by the Plutus Tx compiler.
+Specifically, we can introduce the following function:
+
+```haskell
+nonstrict :: forall a. a -> a
+nonstrict = id
+{-# NOINLINE nonstrict #-}
+```
+
+Then, to create a non-strict let-binding, we write
+
+```haskell
+let x = nonstrict rhs in body
+```
+
+Because of the NOINLINE pragma, `nonstrict` is guaranteed to be visible to the Plutus Tx compiler.
+Whenever the Plutus Tx compiler sees `let x = nonstrict rhs`, it first compiles `rhs` normally; then it creates a nonstrict binding in PIR: `let ~x = rhs_compiled`.
+
+What if `x` gets inlined by GHC, i.e., GHC turns `let x = nonstrict rhs in ...x...` into `...(nonstrict rhs)...`?
+That's not a problem, because if we see a `nonstrict rhs` that isn't the RHS of a binding, we create a new, non-strict binding from it, i.e., it becomes, in PIR,
+
+```haskell
+let ~fresh_name = rhs_compiled in ...fresh_name...
+```
+
+Besides being used to create non-strict let-bindings, the `nonstrict` function can also be used to make it so that an argument to a function is evaluated non-strictly, making it an handy tool for writing Plutus Tx code, especially library code.
+See the next section.
+
+### Using `nonstrict` to Make Arguments Non-strict
+
+Consider the definition of `PlutusTx.foldr` for lists:
+
+```haskell
+foldr f z = go
+  where
+    go = \case
+      []     -> z
+      x : xs -> f x (go xs)
+```
+
+Since Plutus Tx is strict by default, this `foldr` definition cannot short-circuit, because `go xs` will be evaluated before evaluating `f`.
+It is not sufficient to create a non-strict binding for `go xs` (i.e., `let y = go xs in f x y`) because GHC will unconditionally inline `y`.
+
+The problem can be solved by using `nonstrict`, replacing `f x (go xs)` with `f x (nonstrict (go xs))`.
+Now `go xs` is evaluated non-strictly, and `foldr` can short-circuit.
```

Comment:
I think we discussed this but this is not true! It's no good because your non-strict variable then gets used _as a function argument to f_, which means it gets evaluated. You have to _also_ inline `f`, and to allow that you have to first inline `foldr` too.

Any time we have a higher-order function, the use of that function _will_ be strict and we can't easily make it lazy.

Reply:
Oh yeah, that's right, it contradicts the gdrive

Reply:
Oh, right. Then we may need a second magical function, say `nonstrictArg`. Not ideal but it should work. I'll update the PR.

Reply:
No, it still doesn't work. None of these approaches will work in general. _Function application is strict_. There is no way around that, without inlining the function, which requires inlining `foldr`. The problem is not the beta optimization, the problem is _function application_. Seriously, work through the example. I'll pay attention if you can show it works :p

Reply:
Here's how it works: `f x (go xs)` normally becomes `let !acc = go_xs_compiled in f x acc` as a result of beta reduction.
But with `f x (nonstrict (go xs))`, it will annotate `go_xs_compiled` with `NONSTRICT`, and because of that annotation, beta-reduction, will create a nonstrict binding from it, i.e., `let ~acc = go_xs_compiled in f x acc`. Does that make sense?

Reply:
Okay, but now you have to evaluate the call to `f` (e.g. `&&` in the case of `all`)... _which will evaluate acc_. So you have gained nothing.

Reply:
Hmm, right. Looking at [anyCheap.plc.golden](https://github.com/input-output-hk/plutus/blob/master/plutus-tx-plugin/test/Budget/anyCheap.plc.golden#L23) - it will make a difference if `acc` is nonstrict. But we do have `foldr` inlined here, which I guess is your point.

Reply:
Right, in that example we _have_ inlined `foldr` and _also_ `f`. Then, yes, we have a chance. But that's a pretty heavy requirement. And crucially, that's not a property of  the _argument_ to the function, but a property of the function itself (is it going to get inlined?). Making non-strict function application happen is already hard, and I think can only really be done by inlining the function, and we certainly can't do it by just doing something to the _argument_.